packages/htmlbars-util/lib/template-utils.js

```diff
@@ -36,25 +37,30 @@ Block.prototype.invoke = function(env, blockArguments, self, renderNode, parentS
   if (renderNode.lastResult) {
     renderNode.lastResult.revalidateWith(env, undefined, self, blockArguments, visitor);
   } else {
-    let options = { renderState: new RenderState(renderNode) };
-    let { render, template, blockOptions: { scope } } = this;
-    let shadowScope = scope ? env.hooks.createChildScope(scope) : env.hooks.createFreshScope();
-
-    env.hooks.bindShadowScope(env, parentScope, shadowScope, this.blockOptions.options);
+    this._firstRender(env, blockArguments, self, renderNode, parentScope);
+  }
+};
 
-    if (self !== undefined) {
-      env.hooks.bindSelf(env, shadowScope, self);
-    } else if (this.blockOptions.self !== undefined) {
-      env.hooks.bindSelf(env, shadowScope, this.blockOptions.self);
-    }
+Block.prototype._firstRender = function(env, blockArguments, self, renderNode, parentScope) {
+  let options = { renderState: new RenderState(renderNode) };
+  let { render, template, blockOptions: { scope } } = this;
+  let shadowScope = scope ? env.hooks.createChildScope(scope) : env.hooks.createFreshScope();
 
-    bindBlocks(env, shadowScope, this.blockOptions.yieldTo);
+  env.hooks.bindShadowScope(env, parentScope, shadowScope, this.blockOptions.options);
```

Comment:
Can we call `self` something else, im always nervous when we shadow globals. Easy to screw up.l